### ISSUE TITLE
Complexity of sorting

### ISSUE THREAD

**Issue by DougBoffey** [leios/SoME_Topics#41]
It is well known that it takes O(n log n) complexity to sort dara, but why?  That is what I'd like to demonstrate.

A provisional outline would be:

1. Explain the big-O notation
2. Explain the two parts of sorting, (e.g. as a recipe) i.e. ordering the data (following the recipe)  and working out what to do (creating the recipe).  These two operations are somewhat interleaved in any standard sort
3. Show that, once the recipe has been created, it takes O(n) (e.g. sway the first element with whatever, then the second, and so forth)
4. Show it takes O(n log n) to create the recipe
4.1. There are n! ways of ordering n items, so need to distinguish between n! scenarios
4.2. Each comparison splits the scenario space in at best half, so would need at least log[2](n!) operations.
4.3. log[2](n!) = log[2](2 x 3 x ... x n) = log[2](2) + log[2](3) + ... + log[2](n)
4.4. Using log[2](i) x 1 rectangles, show that this is at least integral(log[2](x), x=1 ... n)
4.5. The result follows easily

**Comment by noamtashma** [comment]
It might be worth noting that sorting can in fact be done in O(n) time, in common cases. Sorting only requires O(n log n) if you can only use comparisons.

For example, if you're sorting integers that don't have too many digits, radix sort is O(n).

**Comment by DougBoffey** [comment]
Yes and no.  Basically, radix sorts simply change the base of the logarithm.  Take, for example, lexicographic sorting of words, the logarithm would be to the base of 256.  This does not change the complexity, just the constant multiplier.

**Comment by alexanderskulikov** [comment]
I agree that it is worth noting that the $\Omega(n \log n)$ lower bound applies to comparison-based algorithms only: say, one can easily sort $n$ bits in time $O(n)$.

**Comment by follymath** [comment]
Ah, but don't forget to at least acknowledge [spaghetti sort](https://en.wikipedia.org/wiki/Spaghetti_sort), which is in fact an `O(n)` sort.
